Sources_v3/StreamChat/ChatClient.swift

```diff
@@ -259,59 +272,234 @@ public class _ChatClient<ExtraData: ExtraDataTypes> {
-        log.assert(config.tokenProvider != nil, "You can't call `refreshToken` when `Config.tokenProvider` is nil.")
+}
 
-        guard let currentUserId = currentUserId else {
-            completion(ClientError.CurrentUserDoesNotExist())
-            return
-        }
-        
-        config.tokenProvider?(config.apiKey, currentUserId, { [weak self] (token) in
-            guard let token = token else {
-                completion(ClientError.MissingToken("Can't connect because `TokenProvider` didn't return a valid token."))
-                return
+private extension _ChatClient {
+    func fetchCurrentUserIdFromDatabase() -> UserId? {
+        var currentUser: UserId?
+
+        let context = databaseContainer.viewContext
+        if Thread.isMainThread {
+            currentUser = context.currentUser()?.user.id
+        } else {
+            context.performAndWait {
+                currentUser = context.currentUser()?.user.id
             }
-            self?.currentToken = token
-            completion(nil)
-        })
+        }
+
+        return currentUser
     }
-    
+
     func createBackgroundWorkers() {
         backgroundWorkers = workerBuilders.map { builder in
             builder(self.databaseContainer, self.apiClient)
         } + eventWorkerBuilders.map { builder in
             builder(self.databaseContainer, self.eventNotificationCenter, self.apiClient)
         }
     }
-    
+
+    func completeConnectionIdWaiters(passing connectionId: String?) {
+        _connectionIdWaiters.mutate { waiters in
+            waiters.forEach { $0(connectionId) }
+            waiters.removeAll()
+        }
+    }
+
+    func completeTokenWaiters(passing token: Token?) {
+        _tokenWaiters.mutate { waiters in
+            waiters.forEach { $0(token) }
+            waiters.removeAll()
+        }
+    }
+
     /// Before CIS-555, `webSocketClient` was initialized in `createBackgroundWorkers` in `init` since it was passed
     /// to the background workers. Now, background workers do not access `webSocketClient` and it's not being
     /// initialized in `init`, so we initialize it here explicitly.
-    private func createWebSocketClient() {
+    func createWebSocketClient() {
         _ = webSocketClient
     }
+
+    func prepareEnvironment(newToken: Token, completion: @escaping (Error?) -> Void) {
+        // Check token is for current user.
+        guard newToken.userId != currentUserId else {
+            // Update the current token with the new one.
+            currentToken = newToken
+
+            // Forward the new token to waiting requests.
+            completeTokenWaiters(passing: newToken)
+
+            // It makes more sense to create background workers here
+            // other than in `init` because workers without currently logged-in
+            // user do nothing.
+            if backgroundWorkers.isEmpty {
+                createBackgroundWorkers()
+            }
+
+            completion(nil)
+            return
```

Comment:
Do we also need to recreate the WS connect endpoint here to reflect the new token? Or maybe even disconnect the WS?

Reply:
>Do we also need to recreate the WS connect endpoint here to reflect the new token

no, since the `webSocketConnect` endpoint requires `userId` which's not changed.

> Or maybe even disconnect the WS?

that is a good question 👍 Looks like we should because the previous token could expire